src/textual/widgets/_progress_bar.py

```diff
@@ -0,0 +1,432 @@
+
+        Example:
+            ```py
+            class MyApp(App):
+                def compose(self):
+                    yield ProgressBar(total=100)
+
+                def key_space(self):
+                    self.query_one(ProgressBar).advance()
+            ```
+
+        Args:
+            total: The total number of steps in the progress if known.
+            hide_bar: Whether to hide the bar portion from the progress bar.
+            hide_percentage: Whether to hide the percentage status from the bar.
+            hide_eta: Whether to hide the ETA countdown from the progress bar.
+            name: The name of the widget.
+            id: The ID of the widget in the DOM.
+            classes: The CSS classes for the widget.
+            disabled: Whether the widget is disabled or not.
+        """
+        super().__init__(name=name, id=id, classes=classes, disabled=disabled)
+        self._bar = Bar()
+        self._percentage_status = PercentageStatus()
+        self._eta_status = ETAStatus()
+        self.hide_bar = hide_bar
+        self.hide_percentage = hide_percentage
+        self.hide_eta = hide_eta
+
+        self._percentage = None
+        self.total = total
+
+    def compose(self) -> ComposeResult:
+        with Horizontal():
+            if not self.hide_bar:
+                bar = Bar()
+                # Notify the bar when the percentage is updated.
+                self.watch(
+                    self, "_percentage", partial(setattr, bar, "_completion_percentage")
+                )
+                yield bar
+            if not self.hide_percentage:
+                percentage_status = PercentageStatus()
+                # Notify the percentage status label when the percentage is updated.
+                self.watch(
+                    self,
+                    "_percentage",
+                    partial(setattr, percentage_status, "_completion_percentage"),
+                )
+                yield percentage_status
+            if not self.hide_eta:
+                eta_status = ETAStatus()
+                # Notify the ETA status label when the percentage is updated.
+                self.watch(
+                    self,
+                    "_percentage",
+                    partial(setattr, eta_status, "_completion_percentage"),
+                )
+                yield eta_status
+
+    def validate_progress(self, progress: float) -> float:
+        """Prevent the progress to go over the total number of steps."""
+        if self.total is not None:
+            return min(progress, self.total)
+        return progress
+
+    def watch_total(self, total: float | None) -> None:
+        """Post the message [`ProgressBar.Started`][textual.widgets.ProgressBar.Started]."""
+        if total is None:
+            return
+        if not self._started:
+            self.post_message(ProgressBar.Started(self, total))
+            self._started = True
+
+    def compute__percentage(self) -> float | None:
```

Comment:
Confused about `_percentage` and `percentage` which appear to calculate the same thing?